### PR TITLE
ecds: rationalize ownership

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Commit Message: Rationalize ownership of config providers and subscriptions.
Additional Description: This should fix issue https://github.com/envoyproxy/envoy/issues/14930.

The root cause is unclear ownership of config between listeners and subscriptions (which are shared and operate separately).
The bug is triggered when multiple churning listeners reference the same ECDS resource, while at the same time ECDS pushes updates to it.

This PR fixes most issues discovered:
* Ensure that the config subscription only references the server factory context, not the listener context, as the listener context may go away while the subscription is active.
* Ensure that the asynchronous `onConfigUpdate` call on a worker does not reference the  dynamic config provider as it is owned by a listener and may go away after `onConfigUpdate` but before all workers process the event. The solution is to create a shared pointer intermediary that can outlast the dynamic config provider.
* Ensure that the filter factory is always deleted on main. Regular thread local destructor does not guarantee that. The solution is to issue an empty update to TLS during the destructor. This requires exposing the `isShutdown` accessor to avoid an assertion failure during process shutdown.
* Move the terminal filter detection to the provider from the subscriptiion. This is because the terminal filter condition on a factory depends on the listener context (unlikely to be used, but is a public filter interface).

Risk Level: medium
Testing: black box integration test over long time with racey xDS, unit, added integration test
Docs Changes: none
Release Notes: none

### CODE REVIEW

**Comment by lambdai** [comment]
Several high level questions

> * Ensure that the config subscription only references the server factory context, not the listener context, as the listener context may go away while the subscription is active.

From the perspective of lifetime, this is the right move. It's a interface change to the component which may use context. I guess if the compile succeeds it should be ok? There is some stats scope change along with the context. Perhaps it should be reflected in the release log.



> * Ensure that the filter factory is always deleted on main. Regular thread local destructor does not guarantee that. The solution is to issue an empty update to TLS during the destructor. This requires exposing the `isShutdown` accessor to avoid an assertion failure during process shutdown.

isShutdown seems right in this case. However, there is another corner case that is unrelated to shutdown:
The TLS can be destroyed before the lambda that executing on each worker thread. Is the filter factory deleted on the main thread in this case?

**Comment by kyessenov** [comment]
> From the perspective of lifetime, this is the right move. It's a interface change to the component which may use context. I guess if the compile succeeds it should be ok? There is some stats scope change along with the context. Perhaps it should be reflected in the release log.

Would it be fair to add that ECDS stats now bind to server scope so are not cleared when listeners are removed?

> isShutdown seems right in this case. However, there is another corner case that is unrelated to shutdown: The TLS can be destroyed before the lambda that executing on each worker thread. Is the filter factory deleted on the main thread in this case?

`runOnAllThreads` ensures that the last call is always on main by posting to main from last deletion of shared pointer on a worker. Because we pass the main config as part of the call, it shouldn't matter whether backing TLS is gone?

**Comment by lambdai** [comment]
> Would it be fair to add that ECDS stats now bind to server scope so are not cleared when listeners are removed?
> 

I don't have the answer. 
Server scope is easier to implemented correctly. 
Per listener scope is probably what user wants to see.
I don't understand which is chosen in this PR right now...

**Comment by kyessenov** [comment]
> > Would it be fair to add that ECDS stats now bind to server scope so are not cleared when listeners are removed?
> 
> I don't have the answer. Server scope is easier to implemented correctly. Per listener scope is probably what user wants to see. I don't understand which is chosen in this PR right now...

Scopes are derived from server factory context in this PR, so it's not tied to listener one. However, we still carry on stats_prefix from the listener even when listener changes. I suppose we can changes that so that the stat prefix for ECDS is completely decoupled from listeners.

**Comment by lambdai** [comment]
> > isShutdown seems right in this case. However, there is another corner case that is unrelated to shutdown: The TLS can be destroyed before the lambda that executing on each worker thread. Is the filter factory deleted on the main thread in this case?
> 
> `runOnAllThreads` ensures that the last call is always on main by posting to main from last deletion of shared pointer on a worker. Because we pass the main config as part of the call, it shouldn't matter whether backing TLS is gone?

You are right. The on-main-thread callback is not associated with the slot. Sorry for the churn

**Comment by adisuissa** [comment]
Thanks for working on this @kyessenov!
I'm still looking through the changes and trying to get a better understanding of the subtleties of ECDS, and will give my comments tomorrow.

BTW: I think that this PR also fixes #19628.

**Comment by kyessenov** [comment]
Gentle ping for a review. This is destabilizing Wasm APIs in Istio.

File: source/common/filter/config_discovery_impl.h [hunk 1]
```diff
@@ -125,17 +124,41 @@ class DynamicFilterConfigProviderImpl : public DynamicFilterConfigProviderImplBa
...
-  absl::optional<FactoryCb> current_config_{absl::nullopt};
+  // it. Filter factories may hold their own thread local storage which is required to be deleted
+  // on the main thread.
+  struct MainConfig {
+    absl::optional<FactoryCb> current_config_{absl::nullopt};
+  };
+
+  const std::string stat_prefix_;
+  Server::Configuration::FactoryContext& factory_context_;
+  std::shared_ptr<MainConfig> main_config_;
```
Comment (adisuissa) [inline source/common/filter/config_discovery_impl.h:159]: I'm not certain that shared_ptr makes a difference in this case.
IIUC the only update is on the pointed object (`current_config_`), which would be overrun whether this is a shared pointer or not.

File: test/integration/extension_discovery_integration_test.cc [hunk 2]
```diff
@@ -731,5 +731,59 @@ TEST_P(ExtensionDiscoveryIntegrationTest, BasicFailTerminalFilterNotAtEndOfFilte
   EXPECT_EQ("500", response->headers().getStatusValue());
 }
 
+// Validate that deleting listeners does not break active ECDS subscription.
+TEST_P(ExtensionDiscoveryIntegrationTest, ReloadBoth) {
```
Comment (adisuissa) [inline test/integration/extension_discovery_integration_test.cc:735]: I understand that this is not part of this PR, but I think that the current integration test doesn't exercise the `onConfigRemoved` path because the test doesn't use delta-xDS.

**Review by adisuissa — COMMENTED** [review]
Thanks for fixing this!
Overall I agree with the change, but I have a question regarding the shared pointer.

File: source/common/filter/config_discovery_impl.h (hunk 1 above)
Comment (kyessenov) [reply source/common/filter/config_discovery_impl.h:159]: During destruction, `main_config` is shared with the `runOnAllThreads` callbacks which guarantees that the last reference is deleted on the main thread. That ensures that the inner config is always deleted on the main thread, which is necessary if the inner config has its own TLS.

File: test/integration/extension_discovery_integration_test.cc (hunk 2 above)
Comment (kyessenov) [reply test/integration/extension_discovery_integration_test.cc:735]: Ack, let's do this separately since this looks like some existing debt.

File: test/integration/extension_discovery_integration_test.cc [hunk 3]
```diff
@@ -731,5 +731,59 @@ TEST_P(ExtensionDiscoveryIntegrationTest, BasicFailTerminalFilterNotAtEndOfFilte
...
+  {
+    sendXdsResponse("foo", "2", allowAllConfig());
+    test_server_->waitForCounterGe("http.config_test.extension_config_discovery.foo.config_reload",
+                                   2);
+    auto response = codec_client_->makeHeaderOnlyRequest(banned_request_headers);
+    ASSERT_TRUE(response->waitForEndStream());
+    ASSERT_TRUE(response->complete());
+    EXPECT_EQ("200", response->headers().getStatusValue());
+  }
+}
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply test/integration/extension_discovery_integration_test.cc:786]: Done.

File: source/common/filter/config_discovery_impl.cc [hunk 4]
```diff
@@ -160,8 +151,7 @@ void FilterConfigSubscription::onConfigUpdate(
     last_config_hash_ = 0;
     last_config_ = nullptr;
     last_type_url_ = "";
-    last_filter_is_terminal_ = false;
-    last_filter_name_ = "";
+    last_factory_name_ = "";
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply source/common/filter/config_discovery_impl.cc:154]: Done.

File: source/common/filter/config_discovery_impl.h (hunk 1 above)
Comment (adisuissa) [reply source/common/filter/config_discovery_impl.h:163]: I see, thanks for clarification.

**Review by adisuissa — APPROVED** [review]
LGTM.
/assign-from @envoyproxy/senior-maintainers

**Comment by kyessenov** [comment]
/retest
(timeout)

**Comment by kyessenov** [comment]
Another gentle ping for a crash fix.

**Review by snowp — APPROVED** [review]
Thanks!
